File: spec/java/src/io/kaitai/struct/specwrite/TestProcessToUser.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.ProcessToUser;

import java.io.FileInputStream;

import org.testng.annotations.Test;

public class TestProcessToUser extends CommonSpec {
    @Test
    public void testProcessToUser() throws Exception {
        // NOTE: unlike the automatic roundtrip test, the `_raw_*` fields are set to `null` in this
        // manual test, so "cheating" by just writing them is impossible

        ProcessToUser r = new ProcessToUser();

        ProcessToUser.JustStr buf1 = new ProcessToUser.JustStr(null, r, r._root());
        buf1.setStr("Hello");
        buf1._check();
        r.setBuf1(buf1);

        r._check();

        byte[] expected = new byte[5];
        try (FileInputStream fis = new FileInputStream(SRC_DIR + "process_rotate.bin")) {
            fis.read(expected);
        }

        byte[] actual = new byte[expected.length];
        try (KaitaiStream io = new ByteBufferKaitaiStream(actual)) {
            r._write(io);
        }

        assertByteArrayEquals(actual, expected);
    }

    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return ProcessToUser.class;
    }

    @Override
    protected String getSrcFilename() {
        return "process_rotate.bin";
    }
}

File: spec/python/specwrite/test_process_to_user.py
import io
import unittest
from kaitaistruct import KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.process_to_user import ProcessToUser

class TestProcessToUser(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestProcessToUser, self).__init__(*args, **kwargs)
        self.struct_class = ProcessToUser
        self.src_filename = 'src/process_rotate.bin'

    def test_process_to_user(self):
        # NOTE: unlike the automatic roundtrip test, the `_raw_*` fields are unset in this
        # manual test, so "cheating" by just writing them is impossible

        r = ProcessToUser()

        buf1 = ProcessToUser.JustStr(None, r, r._root)
        buf1.str = "Hello"
        buf1._check()
        r.buf1 = buf1

        r._check()

        with io.open('src/process_rotate.bin', 'rb') as f:
            expected = f.read(5)
        self.assertEqual(len(expected), 5)

        ks_io = KaitaiStream(io.BytesIO(bytearray(len(expected))))
        r._write(ks_io)

        actual = ks_io.to_byte_array()

        self.assert_byte_array_equal(actual, expected)
